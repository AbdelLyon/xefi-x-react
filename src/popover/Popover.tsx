import type { JSX, ReactNode } from "react";
import type { PopoverProps, PopoverContentProps } from "@heroui/react";
import {
  Popover as PopoverRoot,
  PopoverTrigger,
  PopoverContent,
} from "@heroui/react";
import type { Radius } from "@/types";

export type PropsPopover = {
  trigger: ReactNode;
  children: ReactNode;
  contentClassName?: string;
  popoverContentProps?: PopoverContentProps;
  radius?: Radius;
} & Omit<PopoverProps, "content" | "children">;

export const Popover = ({
  trigger,
  children,
  contentClassName,
  popoverContentProps,
  radius = "sm",
  motionProps = {
    variants: {
      enter: {
        y: 0,
        opacity: 1,
        transition: {
          y: { duration: 0.1 },
          opacity: { duration: 0.15 },
        },
      },
      exit: {
        y: "10%",
        opacity: 0,
        transition: {
          y: { duration: 0 },
          opacity: { duration: 0.1 },
        },
      },
    },
  },
  offset = 10,
  placement = "bottom",
  ...props
}: PropsPopover): JSX.Element => {
  return (
    <PopoverRoot
      motionProps={motionProps}
      offset={offset}
      placement={placement}
      radius={radius}
      isOpen={props.isOpen}
      {...props}
    >
      <PopoverTrigger>{trigger}</PopoverTrigger>
      <PopoverContent className={contentClassName} {...popoverContentProps}>
        {children}
      </PopoverContent>
    </PopoverRoot>
  );
};
