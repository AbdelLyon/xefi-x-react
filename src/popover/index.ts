import "@/index.css";
export { Popover } from "@/popover/Popover";
