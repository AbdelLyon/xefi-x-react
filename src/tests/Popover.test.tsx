import { Popover } from "@/popover";
import { render, screen } from "@testing-library/react";
import { describe, it, expect } from "vitest";

describe("Composant Popover", (): void => {
  const mockTrigger = <button>Cliquez-moi</button>;
  const mockContent = <div key="popover-content">Contenu du popover</div>;

  describe("Snapshots", (): void => {
    it("devrait correspondre au snapshot avec le déclencheur", (): void => {
      const { container } = render(
        <Popover trigger={mockTrigger}>{[mockContent]}</Popover>,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot quand ouvert", (): void => {
      const { container } = render(
        <Popover trigger={mockTrigger} isOpen>
          {[mockContent]}
        </Popover>,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec une classe personnalisée pour le contenu", (): void => {
      const { container } = render(
        <Popover trigger={mockTrigger} isOpen contentClassName="custom-class">
          {[mockContent]}
        </Popover>,
      );
      expect(container).toMatchSnapshot();
    });

    it("devrait correspondre au snapshot avec des props de motion personnalisés", (): void => {
      const customMotionProps = {
        variants: {
          enter: { opacity: 1, transition: { duration: 0.2 } },
          exit: { opacity: 0, transition: { duration: 0.1 } },
        },
        initial: "exit",
        animate: "enter",
      };

      const { container } = render(
        <Popover trigger={mockTrigger} isOpen motionProps={customMotionProps}>
          {[mockContent]}
        </Popover>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe("Rendu et comportement", (): void => {
    it("devrait rendre l'élément déclencheur", (): void => {
      render(<Popover trigger={mockTrigger}>{[mockContent]}</Popover>);
      expect(screen.getByText("Cliquez-moi")).toBeInTheDocument();
    });

    it("devrait rendre le contenu quand ouvert", (): void => {
      render(
        <Popover trigger={mockTrigger} isOpen>
          {[mockContent]}
        </Popover>,
      );
      expect(screen.getByText("Contenu du popover")).toBeInTheDocument();
    });

    it("devrait appliquer la classe du contenu", (): void => {
      render(
        <Popover trigger={mockTrigger} isOpen contentClassName="custom-class">
          {[mockContent]}
        </Popover>,
      );

      const contentElement = screen
        .getByText("Contenu du popover")
        .closest(".custom-class");
      expect(contentElement).toBeTruthy();
    });

    it("devrait transmettre les props additionnels du popover", (): void => {
      render(
        <Popover trigger={mockTrigger} isOpen>
          {[mockContent]}
        </Popover>,
      );

      const triggerButton = screen.getByText("Cliquez-moi");
      expect(triggerButton).toHaveAttribute("aria-expanded", "true");
    });
  });
});
