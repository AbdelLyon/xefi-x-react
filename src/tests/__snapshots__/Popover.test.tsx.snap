// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Composant Popover > Snapshots > devrait correspondre au snapshot avec des props de motion personnalisés 1`] = `
<div
  aria-hidden="true"
>
  <button
    aria-controls="react-aria-:r9:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    Cliquez-moi
  </button>
</div>
`;

exports[`Composant Popover > Snapshots > devrait correspondre au snapshot avec le déclencheur 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    Cliquez-moi
  </button>
</div>
`;

exports[`Composant Popover > Snapshots > devrait correspondre au snapshot avec une classe personnalisée pour le contenu 1`] = `
<div
  aria-hidden="true"
>
  <button
    aria-controls="react-aria-:r5:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    Cliquez-moi
  </button>
</div>
`;

exports[`Composant Popover > Snapshots > devrait correspondre au snapshot quand ouvert 1`] = `
<div
  aria-hidden="true"
>
  <button
    aria-controls="react-aria-:r1:"
    aria-expanded="true"
    aria-haspopup="dialog"
    class="z-10 aria-expanded:scale-[0.97] aria-expanded:opacity-70 subpixel-antialiased"
    data-slot="trigger"
    tabindex="0"
    type="button"
  >
    Cliquez-moi
  </button>
</div>
`;
